### PR TITLE
Bugfix - Relative engine association

### PR DESCRIPTION
**Pull request by joshuahuburn**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Added support for paths as the engine association.

#### Release note
Bugfix: Fixed support for relative paths as the engine association in your games .uproject file.

#### Tests
Tested with no path, relative path and registry key.

#### Primary reviewers
@m-samiec @jared-improbable

### CODE REVIEW

File: SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Build/Build.cs [hunk 1]
```diff
@@ -133,6 +139,8 @@ public static void Main(string[] args)
                 Console.WriteLine("Engine is at: " + unrealEngine);
             }
 
+            return;
```
Comment (m-samiec) [inline SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Build/Build.cs]: Oh dear

**Comment by m-samiec** [comment]
Increment requiresSetup

File: SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Build/Build.cs (hunk 1 above)
Comment (jared-improbable) [reply SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Build/Build.cs]: 😱

File: SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Build/Build.cs [hunk 2]
```diff
@@ -89,17 +89,33 @@ public static void Main(string[] args)
...
             string unrealEngine = "";
+            string engineAssociation = "";
 
-            // If we found an engine association within the .uproject, query the registry for the engine directory.
-            if (engineAssociationMatch.Success)
-            {
-                string engineAssociation = engineAssociationMatch.Value;
-                Console.WriteLine("Engine Association: " + engineAssociation);
+            dynamic projectJson = JObject.Parse(uproject);
+            dynamic assocation = projectJson.EngineAssociation;
```
Comment (jared-improbable) [inline SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Build/Build.cs]: I think you can collapse the assignments down even more

File: SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Build/Build.cs [hunk 3]
```diff
@@ -89,17 +89,33 @@ public static void Main(string[] args)
...
+            Console.WriteLine("Engine Association: " + engineAssociation);
 
+            // If the engine association is empty then climb the parent directories of the project looking for the "Engine" and Build.version.
+            if (string.IsNullOrEmpty(engineAssociation))
+            {
+                DirectoryInfo currentDir = new DirectoryInfo(Directory.GetCurrentDirectory());
+                while (currentDir.Parent != null && string.IsNullOrEmpty(unrealEngine))
+                {
+                    currentDir = currentDir.Parent;
+                    unrealEngine = currentDir.GetDirectories().Where(d => d.Name == "Engine" && File.Exists(Path.Combine(d.FullName, @"Build\Build.version"))).Select(d => d.Parent.FullName).FirstOrDefault();
```
Comment (jared-improbable) [inline SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Build/Build.cs]: Why not just build the full path up with all of the information you have and check `File.Exists` on that?
```
var testEngine = Path.Combine(currentDir.FullName, "Engine");
if (File.Exists(testEngine, "Build", "Build.version"))
{
	unrealEngine = testEngine;
	break;
}
```

File: SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Build/Build.cs (hunk 2 above)
Comment (joshuahuburn) [reply SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Build/Build.cs]: True!

File: SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Build/Build.cs (hunk 3 above)
Comment (joshuahuburn) [reply SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Build/Build.cs]: Works too, makes the code bit simpler too.

File: SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Build/Build.cs [hunk 4]
```diff
@@ -89,17 +89,37 @@ public static void Main(string[] args)
             // Locate the Unreal Engine.
             Console.WriteLine("Finding Unreal Engine build.");
             string uproject = File.ReadAllText(projectFile, Encoding.UTF8);
-            string engineAssociationRegex = "..{8}-.{4}-.{4}-.{4}-.{12}.";
-            Match engineAssociationMatch = Regex.Match(uproject, engineAssociationRegex);
 
             string unrealEngine = "";
+            string engineAssociation = "";
```
Comment (m-samiec) [inline SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Build/Build.cs]: Might as well assign this directly to the projectJson version.

Comment (joshuahuburn) [reply SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Build/Build.cs]: Done

File: SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Build/Build.cs [hunk 5]
```diff
@@ -89,17 +89,36 @@ public static void Main(string[] args)
...
-            // If we found an engine association within the .uproject, query the registry for the engine directory.
-            if (engineAssociationMatch.Success)
+            // If the engine association is empty then climb the parent directories of the project looking for the Unreal Engine root directory.
+            if (string.IsNullOrEmpty(engineAssociation))
             {
-                string engineAssociation = engineAssociationMatch.Value;
-                Console.WriteLine("Engine Association: " + engineAssociation);
-
+                DirectoryInfo currentDir = new DirectoryInfo(Directory.GetCurrentDirectory());
+                while (currentDir.Parent != null && string.IsNullOrEmpty(unrealEngine))
```
Comment (jared-improbable) [inline SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Build/Build.cs]: The `IsNullOrEmpty` test is redundant given the `break` below

Comment (joshuahuburn) [reply SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/Build/Build.cs]: Good point

**Pull request merged**